src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs

```diff
@@ -90,8 +90,7 @@ pub(super) fn relate_type_and_user_type<'tcx>(
         NllTypeRelatingDelegate::new(infcx, borrowck_context, locations, category),
         v1,
         b_variables,
-    ).relate(&b_value, &a)?;
-    Ok(())
+    ).relate(&b_value, &a).map_err(From::from)
```

Comment:
This is indeed not *quite* what I had in mind. I was imagining something more like this. 

```rust
let mut relating = TypeRelating::new(...);
relating.relate(&b_value, &a)?;

let canonical_var_values = CanonicalVarValues {
    var_values: relating.canonical_var_values
};

Ok(b.substitute(&canonical_var_values))
```

what this would do is to take the canonical type and replace all the "canonical variables" with the values that we inferred during `relate_tys`. It would not require the changes to "propagate back" a correct type. Those make me a bit nervous, as it's not really obvious what to return for `a <: b`. And as a case in point, in the case of handling higher-ranked things, we convert `a == b` to `a <: b` and `b <: a`, so which return value should we give back? etc